### PR TITLE
feat: implement basic header modifier base on service worker

### PR DESCRIPTION
**Pull request by Asuka109**
### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at c8ee6c0</samp>

This pull request adds a new feature to the devtools client and plugin for Modern.js, which allows users to modify request headers using a service worker script. It adds a new webpack plugin, a new UI tab, a new state module, and several new components and utils for the header modifier feature. It also refactors the rpc module and updates the package.json and tsconfig.json files. It creates a new service-worker.ts file, which defines the logic of the header modification service worker. It also adds a new cli.ts file, which sets up the server route and the HTML injection for the service worker script.

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at c8ee6c0</samp>

*  Add a header modifier feature to the devtools client, which allows modifying the headers of network requests using a service worker script and a UI tab. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-c454da44af4a96fc4c1f2154909c6433a87bf21d28eede9789a530a06f1146bcL1-R5), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-249610e4c3d6bd8dbb483f2667c5093d38461da73b5a217a63a55cbe74e0f2a3R4), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-249610e4c3d6bd8dbb483f2667c5093d38461da73b5a217a63a55cbe74e0f2a3R52-R54), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-ee0bdc3f92472c79376377721c0e8c205228ee40cc94ffd50d4feb71deba0c2eL25-R27), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-ee0bdc3f92472c79376377721c0e8c205228ee40cc94ffd50d4feb71deba0c2eL36-R39), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-ee0bdc3f92472c79376377721c0e8c205228ee40cc94ffd50d4feb71deba0c2eR45), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-ee0bdc3f92472c79376377721c0e8c205228ee40cc94ffd50d4feb71deba0c2eR69-R71), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-c1e7d00fa8cea23d41d1fe19190c9c01389dde0c6c6b21f582a91c4c050ac569R1-R41), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-e323dc4c0e3bfbdd28c99cdcc46d867722e3f65089f1f3499ba62100d6ec9967R6), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-e323dc4c0e3bfbdd28c99cdcc46d867722e3f65089f1f3499ba62100d6ec9967R35-R40), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-9108f8ea29e4d8fe62f7cbdae9ad9149496c2b4e201b4af2c6cf4b876bfe31b0R1-R50), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-0e272ce69a3adea7bf625c806870b7e57f6bbae850b2d009b55eafe4eb7b75b3R1-R17), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-11165ab4beffb8ba1c492792c5d2ba4e20e38c307986123ce17662bd554068b2R1-R47), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-7b84d6ec6cfb2d8d758e9191ee16df8f5f50e31e3da8fa85e8db429f56875dc5R1-R25), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-07e6196ea9ff025b89c9a3ce118de74c02f4e4b0a1b4dfe9b5b53a1aef1a3c27R1-R11), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-0e8b8820f266c855801d8ed2808b11a97754f1d176c3fde1e4380a1840290839R1-R17), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-9e33cc908d7c4261a87c7ea76a27a521846b435ff17cbdae36f328c11eced4c7R1-R70), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-a8768a8298817554634e14ece83fbd3fbd989609bea7d29b2adaa37f44578cfdR7), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-a8768a8298817554634e14ece83fbd3fbd989609bea7d29b2adaa37f44578cfdL12-R20), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-2f846ce94e6453db4870e9758df4c3b79d2bb5031b231559a139fbb549b54655R45-R53), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-2f846ce94e6453db4870e9758df4c3b79d2bb5031b231559a139fbb549b54655R65-R68), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-2f846ce94e6453db4870e9758df4c3b79d2bb5031b231559a139fbb549b54655R74-R84))
  * Define the `ServiceWorkerCompilerPlugin` class in `plugins/sw-compiler-plugin.ts`, which creates a child compiler for the service worker script using the `@rspack/core` package and the `DefinePlugin`. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-c1e7d00fa8cea23d41d1fe19190c9c01389dde0c6c6b21f582a91c4c050ac569R1-R41), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-ee0bdc3f92472c79376377721c0e8c205228ee40cc94ffd50d4feb71deba0c2eR45))
  * Add the `ServiceWorkerCompilerPlugin` to the webpack chain in `modern.config.ts`, using the `plugin` and `use` methods. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-249610e4c3d6bd8dbb483f2667c5093d38461da73b5a217a63a55cbe74e0f2a3R4), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-249610e4c3d6bd8dbb483f2667c5093d38461da73b5a217a63a55cbe74e0f2a3R52-R54))
  * Add a new export field to the `package.json` file, which maps the `./sw-proxy` subpath to the `./dist/public/sw-proxy.js` file. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-ee0bdc3f92472c79376377721c0e8c205228ee40cc94ffd50d4feb71deba0c2eL25-R27))
  * Add a new `outDir` property to the `tsconfig.json` file, which specifies the `dist` folder as the output directory for the compiled TypeScript files. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-a8768a8298817554634e14ece83fbd3fbd989609bea7d29b2adaa37f44578cfdR7))
  * Add the `plugins` folder to the `include` array in the `tsconfig.json` file, which specifies the files and folders that are included in the TypeScript compilation. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-a8768a8298817554634e14ece83fbd3fbd989609bea7d29b2adaa37f44578cfdL12-R20))
  * Define the service worker script in `src/service.worker.ts`, which uses the `idb-keyval` package, the `process.env.VERSION` variable, and the `rules` parameter to store and modify the headers of the network requests. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-9e33cc908d7c4261a87c7ea76a27a521846b435ff17cbdae36f328c11eced4c7R1-R70), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-ee0bdc3f92472c79376377721c0e8c205228ee40cc94ffd50d4feb71deba0c2eR69-R71))
  * Define the `handleServiceWorker` function in `src/mount/proxy/index.ts`, which registers and unregisters the service worker script using the `navigator.serviceWorker` API. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-0e8b8820f266c855801d8ed2808b11a97754f1d176c3fde1e4380a1840290839R1-R17))
  * Define the `ModifyHeaderRule` and `ServiceStatus` interfaces in `src/client/utils/service-agent.ts`, which describe the shape of the objects used for the header modification feature. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-07e6196ea9ff025b89c9a3ce118de74c02f4e4b0a1b4dfe9b5b53a1aef1a3c27R1-R11))
  * Define the state for the header modifier feature in `src/client/routes/headers/state.ts`, which uses the `valtio` package to create a reactive state object with a `service` property. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-11165ab4beffb8ba1c492792c5d2ba4e20e38c307986123ce17662bd554068b2R1-R47))
  * Define the `registerService`, `unregisterService`, and `fetchServiceStatus` functions in `src/client/routes/headers/state.ts`, which use the `navigator.serviceWorker` API to register, unregister, and get the status of the service worker script. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-11165ab4beffb8ba1c492792c5d2ba4e20e38c307986123ce17662bd554068b2R1-R47))
  * Define the `Page` component for the header modifier view in `src/client/routes/headers/page.tsx`, which uses the `useSnapshot` and `useNavigate` hooks to check the service status and navigate to the editor or welcome view. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-0e272ce69a3adea7bf625c806870b7e57f6bbae850b2d009b55eafe4eb7b75b3R1-R17))
  * Define the `Page` component for the header modifier welcome view in `src/client/routes/headers/welcome/page.tsx`, which uses the `Link` component to navigate to the editor view and the service worker script. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-7b84d6ec6cfb2d8d758e9191ee16df8f5f50e31e3da8fa85e8db429f56875dc5R1-R25))
  * Define the `Page` component for the header modifier editor view in `src/client/routes/headers/editor/page.tsx`, which uses the `useList` and `useSnapshot` hooks to manage and update an array of rules. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-9108f8ea29e4d8fe62f7cbdae9ad9149496c2b4e201b4af2c6cf4b876bfe31b0R1-R50))
  * Add a new object to the `defaultTabs` array in `src/client/constants.tsx`, which defines the properties of the header modifier tab. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-e323dc4c0e3bfbdd28c99cdcc46d867722e3f65089f1f3499ba62100d6ec9967R35-R40))
  * Import the `HiOutlineAcademicCap` icon from the `@radix-ui/react-icons` package in `src/client/constants.tsx`, which is used for the header modifier tab. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-e323dc4c0e3bfbdd28c99cdcc46d867722e3f65089f1f3499ba62100d6ec9967R6))
  * Add a new method to the `devtoolsPlugin` object in `plugin/src/cli.ts`, which modifies the server routes for the `/sw-proxy.js` path. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-2f846ce94e6453db4870e9758df4c3b79d2bb5031b231559a139fbb549b54655R45-R53))
  * Add new properties to the `webpackConfig` object in `plugin/src/cli.ts`, which copy the service worker script file to the public folder and inject a script tag into the HTML template. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-2f846ce94e6453db4870e9758df4c3b79d2bb5031b231559a139fbb549b54655R65-R68), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-2f846ce94e6453db4870e9758df4c3b79d2bb5031b231559a139fbb549b54655R74-R84))
* Refactor the `rpc` module, which provides a bidirectional RPC communication between the devtools client and the server. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-56e2f60268eb53a6066f89a1511678d450fdcec040909d6bc025bfe57232353fL1-R1), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-09e6ba9f74fbf0329238e05f543a55c8ff41a690dce96d349b130cb1b6ef33ecR1-R51))
  * Move the logic of the `setupServerConnection` function to a separate file, which is `src/client/rpc/server.ts`. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-09e6ba9f74fbf0329238e05f543a55c8ff41a690dce96d349b130cb1b6ef33ecR1-R51))
  * Replace the content of the `src/client/rpc/index.ts` file with a single export statement, which re-exports everything from the `src/client/rpc/server.ts` file. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-56e2f60268eb53a6066f89a1511678d450fdcec040909d6bc025bfe57232353fL1-R1))
* Add a new utility function, which is the `assert` function. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-41a0cd21652a663d91b8ce923b58a5a2ef7063db06060c8b15318f915ff39d60R1-R12), [link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-1ca04e2f9584a05c7d6660083da1bb6c5c4e31c8522a1dd2e366193402f04233R2-R3))
  * Define the `assert` function in `src/client/utils/assert.ts`, which takes a value and an optional message or error, and throws an error if the value is falsy. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-41a0cd21652a663d91b8ce923b58a5a2ef7063db06060c8b15318f915ff39d60R1-R12))
  * Add a new export statement to the `src/client/utils/index.ts` file, which re-exports everything from the `src/client/utils/assert.ts` file. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-1ca04e2f9584a05c7d6660083da1bb6c5c4e31c8522a1dd2e366193402f04233R2-R3))
* Reorder the `devDependencies` alphabetically in the `package.json` file, which is a common convention for organizing dependencies. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-ee0bdc3f92472c79376377721c0e8c205228ee40cc94ffd50d4feb71deba0c2eL36-R39))
* Add a type annotation for the `module.exports` object in the eslint configuration file, which helps with type checking and code completion. ([link](https://github.com/web-infra-dev/modern.js/pull/4975/files?diff=unified&w=0#diff-c454da44af4a96fc4c1f2154909c6433a87bf21d28eede9789a530a06f1146bcL1-R5))

## Checklist

- [ ] I have added changeset via `pnpm run change`.
- [ ] I have updated the documentation.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: packages/devtools/client/plugins/sw-compiler-plugin.ts
Comment (chenjiahan) [inline packages/devtools/client/plugins/sw-compiler-plugin.ts:1]: file name should be camelCase

File: packages/devtools/client/package.json [hunk 1]
```diff
@@ -33,15 +33,16 @@
...
     "@modern-js/runtime": "workspace:*",
     "@modern-js/tsconfig": "workspace:*",
     "@modern-js/types": "workspace:*",
+    "@modern-js/utils": "workspace:*",
     "@radix-ui/react-collapsible": "^1.0.3",
     "@radix-ui/react-icons": "^1.3.0",
     "@radix-ui/react-popover": "^1.0.7",
     "@radix-ui/react-tabs": "^1.0.4",
     "@radix-ui/themes": "^2.0.0",
+    "@rspack/core": "^0.3.12",
```
Comment (chenjiahan) [inline packages/devtools/client/package.json:45]: can we use same rspack version / instance as other pkgs

File: packages/devtools/client/src/client/routes/headers/editor/page.tsx
Comment (chenjiahan) [inline packages/devtools/client/src/client/routes/headers/editor/page.tsx:1]: Should react component files use PascalCase?

File: packages/devtools/client/plugins/sw-compiler-plugin.ts
Comment (Asuka109) [reply packages/devtools/client/plugins/sw-compiler-plugin.ts:1]: done~

File: packages/devtools/client/src/client/routes/headers/editor/page.tsx
Comment (Asuka109) [reply packages/devtools/client/src/client/routes/headers/editor/page.tsx:1]: Well, it just a conventional route file.

**Review by chenjiahan — APPROVED** [review]

**Pull request merged**